### ISSUE TITLE
Get-NexposeAsset returns different total value based on different ?size value

### ISSUE THREAD

**Issue by naimco** [My-Random-Thoughts/Rapid7Nexpose#13]
Just some observations I use the Get-NexposeAsset to get all our assets without any arguments 

$allAssets = $null
Measure-Command { $allAssets = Get-NexposeAsset }
$allAssets.count 

When comparing the total asset numbers from web I see difference :
InSighVM numbers show a total of 822 assets.
The Get-NexposeAsset returns 722 assets.

So I modified the Invoke-NexposeQuery.ps1 line 84   $iRestM.Uri += '?size='  and here are the different result I got : 

APIMaxResults = 1
Count = 821
Minutes           : 4
Seconds           : 47

APIMaxResults = 2
Count = 821
Minutes           : 4
Seconds           : 29

APIMaxResults = 4
Count = 818
Minutes           : 2
Seconds           : 53

APIMaxResults = 5
Count = 817
Minutes           : 2
Seconds           : 45

APIMaxResults = 10
Count = 812
Minutes           : 2
Seconds           : 43
Milliseconds      : 501

APIMaxResults = 100
Count = 722
Minutes           : 1
Seconds           : 59
Milliseconds      : 566

So I have settled on using the APImaxResult=4 for now to make sure I get 99% of the assets,  Any suggestion on how to figure out why assets are missing as the page ?size= increases ?

I also modified all the (ConvertTo-Json -InputObject $ApiQuery -Depth 1800) I tried various increasing numbers here with no change to the count values

**Comment by My-Random-Thoughts** [comment]
That is strange.  I'll have a look at the code.
Can you see if there is anything special about the ones that are "missed".  For example, are they duplicates or filtered on a particular type/value.

As for the `ConvertTo-Json -Depth` value, setting it to a large number can cause PS to crash.  I have had issues with number over 10 sometimes - it depends on the JSON being converted.

**Comment by naimco** [comment]
I had not realized how much data the raw api is returning , allot.

One item of note, when using the raw api, with a browser :  https://oursite.com:3780/api/3/assets?size=1 or 
https://oursite.com:3780/api/3/assets?size=2000

the return data show the totalResources, which is correct. 

  "page" : {
    "number" : 0,
    "size" : 500,
    "totalResources" : **822**,
    "totalPages" : 2
  },

So the resources must be getting lost in parsing of the Json.

I have found that by exporting the asset list in InsighVM (Assets.csv) this file contains a column called AssetKey,  which is a guid-default-asset-ID,  the ending ID matches the ID returned the nexpose API.  So I  just have compare both exported sets to find which assets are missing, then examine the additional properties they may contain.

**Comment by My-Random-Thoughts** [comment]
If you run the command with `-verbose` it should tell you the same information of total size and number of pages.

**Comment by naimco** [comment]
The verbose setting give this type of output :
VERBOSE: Executing API method "GET" against "/api/3/assets?size=1"
VERBOSE: GET https://xxxxxxxx.com:3780/api/3/assets?size=1 with 0-byte payload
VERBOSE: received -1-byte response of content type application/json;charset=UTF-8
VERBOSE: Retreving page   2 of 825 ...

Today I set the size query to 1  size=1.

When I export all the assets I realize that the API returns them in order of ID,  so its very easy to see which ID are missing. Today it was ID 2.  

Get-NexposeAsset -Id '2' -Verbose  ( Executing API method "GET" against "/api/3/assets/2?size=1") Does return the asset !

I also manually found the asset 2 in the raw ouput of the Get-NexposeAsset  is was on page 2,  (https://xxxx.com:3780/api/3/assets?page=1&size=1&sort=id,asc) 

I saved the content of both pages to two different notepad documents and compared them with winmerg and they are different in multiple ways. I was not expecting this.   I can send you samples if you want to take a look , I just cannot post them here in a public form.  

So now the only way I have to get a valid dataset, is to parse the results of Get-NexposeAsset, setup a counter from 1 to 1000 and look to see if have an ID match foreach, if I don't I can add the missing ID back to my array using get-NexposeAsset -id xx

**Comment by My-Random-Thoughts** [comment]
Looking at your testing above (thank you very much for that), it looks like its missing an entire "page" of results.

APIMaxResults = 1
Count = 821

APIMaxResults = 10
Count = 812

APIMaxResults = 100
Count = 722

Adding the APIMaxResults value to the count gets you your total (within one or two).  I wonder if the code is skipping a page for some reason.  Your saved data should show this.

If you want, send them to `support @ myrandomthoughts.co.uk`

**Comment by My-Random-Thoughts** [comment]
Quick test, can you change the start of line 157 to be `1..$totalPages....`

This could be a quick fix!!

**Comment by naimco** [comment]
Very nice,  Invoke-nexposeQuery.ps1 Line 157 set to  1..$totalPages  and Line 84 back to ?size=100' All the assets are now reported ! 

Thank you very much.

**Comment by My-Random-Thoughts** [comment]
Thank you for confirming, I'll make the change permanent.

**Comment by naimco** [comment]
My mistake issue is still present, sent some sample data to support email.

**Comment by My-Random-Thoughts** [comment]
Thank you for your email, can you try a few things for me please if you don't mind.

Log on to your Nexpose console (to get a session ticket), then change the URL to:
**https://server:3780/api/3/assets?page=0&size=100**

It should return the first 100 entries and at the bottom will be the total size and number of pages.  Save this list somewhere you can refer too please.  Now change the **size=100** to more than the assets you have, **1000** or so.  It should load them all and the page total at the bottom should just say 1.

Change the size to now say **10** and see the result, it should match the first 10 as previously.  Now change to **page=1** and see if that list matches 11-20 of the previous list.  Continue with increasing the page number ensuring the server list matches.

Basically we are just manually doing what the script is doing.  It seems to work OK for me with my limited 28 assets at home, so I am not sure where the issue is.  It may have to wait until Monday when I can test it at work with a few more servers.

**Comment by naimco** [comment]
I tried 3 different size settings and it seem the api caps the size to 500 max.

**https://server:3780/api/3/assets?page=0&size=100**
  "page" : {
    "number" : 0,
    "size" : 100,
    "totalResources" : 828,
    "totalPages" : 9
  },
  
  **https://server:3780/api/3/assets?page=0&size=1000**
    "page" : {
    "number" : 0,
    "size" : 500,
    "totalResources" : 828,
    "totalPages" : 2
    
   **https://server:3780/api/3/assets?page=0&size=1500**   
    "page" : {
    "number" : 0,
    "size" : 500,
    "totalResources" : 828,
    "totalPages" : 2
    
    I will do the rest of the tests you suggested  and report back on Monday.
    
    Thanks

**Comment by My-Random-Thoughts** [comment]
I have done some testing with my POC server at work and I was getting some strange results, however that was due to having duplicate IPs with different host names.  I removed all entries and scanned everything again

Once complete, the web GUI, postman (url tester for tests I asked you to test), and the module all agreed on the same values.

I'll leave this open for you to complete any testing and if nothing comes up I'll close it at the end of the week.  Do let me know of any issues though

**Comment by SeanQuinlan** [comment]
I believe the problem is with the paging when using POST in Invoke-NexposeQuery, specifically these lines:

```
If ($_ -eq 2) {
    $iRestM.Uri += '&page=2'
}
Else {
    $iRestM.Uri = $iRestM.Uri.Replace("&page=$($_ - 1)", "&page=$_")
}
```

If you add a Write-Verbose block after the above, you can see the URL does not get the correct substitution, so it ends up requesting the same page over and over again.

Here's the output from me trying to get a list of 299 assets in a site:

```
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=0"
VERBOSE: POST https://127.0.0.1:3780/api/3/assets/search?size=5&page=0 with -1-byte payload
VERBOSE: received -1-byte response of content type application/json;charset=UTF-8
VERBOSE: Removing "links" item from results
VERBOSE: Retreving page  1 of 59 ...
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=1"
VERBOSE: Removing "links" item from results
VERBOSE: Retreving page  2 of 59 ...
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=1&page=2"
VERBOSE: Removing "links" item from results
VERBOSE: Retreving page  3 of 59 ...
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=1&page=3"
VERBOSE: Removing "links" item from results
VERBOSE: Retreving page  4 of 59 ...
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=1&page=4"
VERBOSE: Removing "links" item from results
VERBOSE: Retreving page  5 of 59 ...
VERBOSE: Executing API method "POST" against "/api/3/assets/search?size=5&page=1&page=5"
VERBOSE: Removing "links" item from results
```

It gets page 0, then page 1, then just keeps getting page 1 over and over (up to totalpages), as the API will interpret the first &page=X as the page you're requesting.

You can also see it from the results, where the number of results always equals: totalpages x 5

For my site of 299 assets, it was returning 300 assets. But when I sorted them by hostname and removed duplicates, there were only 10 unique values as expected (page 0 + page 1).

Anyway, I submitted a pull request to fix this (https://github.com/My-Random-Thoughts/Rapid7Nexpose/pull/14), by adding a regex check for the relevant page string and only adding the page=1 when needed. I haven't done extensive testing with this update, but it worked fine for getting all sites and getting all assets in a site.

**Comment by My-Random-Thoughts** [comment]
Thank you very much Sean, I'll take a look

**Comment by My-Random-Thoughts** [comment]
Pull request approved and merged, thank you.

**Comment by naimco** [comment]
Thank you very much for figuring out the bug and fixing it.  Get-NexposeAssets is now returning all our assets.
